unified-runtime/source/adapters/native_cpu/enqueue.cpp

```diff
@@ -692,3 +692,54 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueNativeCommandExp(
     const ur_event_handle_t *, ur_event_handle_t *) {
   return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
 }
+
+UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunchWithArgsExp(
+    ur_queue_handle_t hQueue, ur_kernel_handle_t hKernel,
+    const size_t pGlobalWorkOffset[3], const size_t pGlobalWorkSize[3],
+    const size_t pLocalWorkSize[3], uint32_t numArgs,
+    const ur_exp_kernel_arg_properties_t *pArgs,
+    uint32_t numPropsInLaunchPropList,
+    const ur_kernel_launch_property_t *launchPropList,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    ur_event_handle_t *phEvent) {
+  for (uint32_t argIndex = 0; argIndex < numArgs; argIndex++) {
+    switch (pArgs[argIndex].type) {
+    case UR_EXP_KERNEL_ARG_TYPE_VALUE:
+      hKernel->addArg(pArgs[argIndex].value.value, pArgs[argIndex].index,
+                      pArgs[argIndex].size);
```

Comment:
This duplicates the logic of functions such as `urKernelSetArgValue` except with subtly different error checking, which makes things a bit difficult maintenance-wise. Would it be possible to instead call `urKernelSetArgValue` instead (and make sure it gets inlined so that performance is not affected)? Likewise for the other cases below.

Reply:
I've moved the logic and arg validation into the kernel member functions. The only thing this leaves out is null checking the kernel handle, I'd suggest this isn't needed as it should either happen in the validation layer or not at all, but I could add a null check at the top of the `LaunchWithArgs` implementation. I was reluctant to just call the `urKernelSetArg` functions from here because we can't inline them (doing so seems to mess with linking the interface loader), but I'll make that change without the inline if you don't like this attempt at a re-design